src/model/document.ts

```diff
@@ -290,7 +290,11 @@ export class Document implements Feature {
     }
     warnings.push.apply(warnings, this.warnings);
     for (const feature of this.getFeatures(deep)) {
-      warnings.push.apply(warnings, feature.warnings);
+      for (const warning of feature.warnings) {
+        if (warnings.indexOf(warning) === -1) {
```

Comment:
Hm, this is O(n^2) in the number of warnings. Consider either making `warnings` a Set or maybe check whether `feature === this`